File: Algorithms/src/Number_2/Merge.java
package Number_2;

/**
 * P171 自顶向下的归并排序
 * 
 * @author he
 *
 */
public class Merge {
	private static Comparable aux[];

	public static void sort(Comparable a[]) {
		aux = new Comparable[a.length];
		sort(a, 0, a.length - 1);

	}

	private static void sort(Comparable a[], int firstIndex, int lastIndex) {
		if (firstIndex>=lastIndex) {
			return;
		}
		int midIndex=firstIndex+(lastIndex-firstIndex)/2;
		sort(a,firstIndex,midIndex);//左边排序
		sort(a, midIndex+1, lastIndex);//右边排序
		merge(a, firstIndex, midIndex, lastIndex);//归并结果
	}

	private static void merge(Comparable a[], int firstIndex, int midIndex, int lastIndex) {
		int i = firstIndex;
		int j = midIndex + 1;
		// 将数组a里的元素复制到aux中
		for (int k = firstIndex; k <= lastIndex; k++) {
			aux[k] = a[k];
		}

		for (int k = firstIndex; k <= lastIndex; k++) {
			if (i > midIndex) {
				a[k] = aux[j++];// 左边取完取右边
			} else if (j > lastIndex) {
				a[k] = aux[i++];// 右边取完取左边
			} else if (aux[j].compareTo(aux[i]) < 0) {
				a[k] = aux[j++];
			} else {
				a[k] = aux[i++];
			}
		}
	}
	
	public static void show(Comparable a[]) {
		int N = a.length;
		for (int i = 0; i < N; i++) {
			System.out.print(a[i] + " ");
		}
	}
	
	public static void main(String[] args) {
		String a[] = { "E", "E", "G", "M", "R", "A", "E", "C", "R", "T" };
		sort(a);
		show(a);
	}

}

File: AlgorithmsByKotlin/src/Number_2/Merge.kt
package Number_2

/**
 *  归并排序
 * Created by he on 2017/5/28.
 */

class Merge{
 lateinit var aux: IntArray
    fun sort(a: Array<Int>) {
        aux = IntArray(a.size)
        sort(a, 0, a.size - 1)
    }

    fun sort(a: Array<Int>, lo: Int, hi: Int) {
        if (lo >= hi)
            return
        var mid = lo + (hi - lo) / 2
        sort(a, lo, mid)
        sort(a, mid + 1, hi)
        merge(a, lo, hi, mid)
    }

    fun merge(a: Array<Int>, lo: Int, hi: Int, mid: Int) {
        var i = lo
        var j = mid + 1
        for (k in lo..hi) {
            aux[k] = a[k]
        }

        for (k in lo..hi) {
            when {
                i > mid -> a[k] = aux[j++]
                j > hi -> a[k] = aux[i++]
                aux[j] < aux[i] -> a[k] = aux[j++]
                else -> a[k] = aux[i++]
            }
        }

    }
}

fun main(args: Array<String>) {
    var a = arrayOf(1, 5, 2, 68, 5, 2, 10, 4)
    val  merge=Merge()
    merge.sort(a)
    show(a)
}




